### ISSUE TITLE
Add analytics tracking

### ISSUE THREAD

**Issue by kstaken** [terascope/teraserver#52]
We need to get a little better insight into the query volume when this service is deployed.

We'll want to track the following items.

     * all requests
          * count by status code
          * query time 
               * average over 10 seconds?

     * per token - truncated to first 5 chars
          * count by status code
          * query time 
               * average over 10 seconds?

This should just be aggregated in memory and dumped to Elasticsearch every 10 seconds. 

One critical requirement is that this analytics process should NEVER impact the ability of the service to respond to API calls. It is better to lose analytics temporarily than to have API responses impacted.

**Comment by kstaken** [comment]
Here's a module we use for something similar in other scenarios.

```
'use strict';

const _ = require('lodash');

/**
 * Configuration
 * datatype
 * module
 * name
 * connection
 * cluster
 */
module.exports = function module(context, statsConfig) {
    const config = context.sysconfig;
    const logger = context.logger;

    // If we aren't running in a context where this is available just
    // return a mock.
    if (!context.foundation || !context.foundation.getConnection) {
        return {
            count() {}
        };
    }

    const esStats = context.foundation.getConnection({
        endpoint: statsConfig.connection,
        type: 'elasticsearch',
        cached: true
    }).client;

    if (!esStats) {
        throw new Error(`Unknown elasticsearch connection for stats ${statsConfig.connection}`);
    }

    let counters = {};

    function internalIncrement(counter, increment) {
        let realIncrement = increment;
        if (realIncrement === undefined || realIncrement === null) {
            realIncrement = 1;
        }

        if (!Object.prototype.hasOwnProperty.call(counters, counter)) { counters[counter] = 0; }

        counters[counter] += realIncrement;
    }

    function formattedDate(date) {
        return date.slice(0, 7).replace(/-/gi, '.');
    }

    // Default to flushing to ES every 10 seconds
    const flushInterval = 10000; // ms

    setInterval(() => {
        const sending = counters;
        counters = {};

        let nodename = config._nodeName.split('.');
        const workerId = nodename.pop();
        nodename = nodename.join('.');

        const bulkRequest = [];
        _.forOwn(sending, (value, counter) => {
            const timestamp = new Date();
            const syncDate = formattedDate(timestamp.toISOString(), true);

            bulkRequest.push({
                index: {
                    _index: `counters-${syncDate}`,
                    _type: 'counter'
                }
            });

            const record = {
                date: timestamp,
                cluster: statsConfig.cluster,
                service: statsConfig.service,
                node: nodename,
                worker: workerId,
                module: statsConfig.module,
                counter,
                count: value
            };

            if (statsConfig.datatype) { record.datatype = statsConfig.datatype; }

            bulkRequest.push(record);
        });

        if (bulkRequest.length > 0) {
            flush(bulkRequest);
        }
    }, flushInterval);

    function flush(bulkSave) {
        // logger.info("Flushing " + bulkSave.length + " stats records to elasticsearch.");
        esStats.bulk({
            body: bulkSave
        })
            .then((resp) => {
                resp.items.forEach((item) => {
                    if (item.create && item.create.status !== 201) {
                        logger.error(`Stats sync error from bulk insert: ${item.create.status} ${item.create.error}`);
                    }
                });
            })
            .catch((err) => {
                logger.error(`Error syncing stats data to ES: ${err}`);
            });
    }

    return {
        count(counter, increment) {
            internalIncrement(counter, increment);
        }
    };
};
```

**Comment by ciorg** [comment]
here's the timer in lib/search.js line 
```
    try {
            const start = Date.now();
            performSearch({index: index, ignoreUnavailable: true}, req, res, config);
            const end = Date.now();
            res.apiSearchTime = end - start;
        }
```

idea for the data structure

```
countObject = { token1: { statusCode: count, 
                                           statusCode: count, 
                                            etc...,
                                           apiSearchTimes: [time1, time2, time3, etc...]},
                             token2: {{ statusCode: count, 
                                           statusCode: count, 
                                            etc...,
                                           apiSearchTimes: [time1, time2, time3, etc...]},
                            }
```

Not sure where to call the counter/flush function.  I started calling it in the logscope/ index.js, but now I'm thinking that won't work because other plugins will also use the search function.  Will need to figure that out and I'm still hacking on how to start the process and how the flush will work together.

**Comment by ciorg** [comment]
also here' the code I have to keep the count
```
const statusCode = res.statusCode;
    const userToken = req.user.api_token;
    const searchTime = res.apiSearchTime;

    if (tally[userToken] && tally[token][statusCode]) {
        tally[userToken][statusCode] += 1;
    } else if (tally[userToken]){
        tally[userToken][statusCode] = 1;
    } else {
        tally[userToken] = {};
        tally[userToken][statusCode] = 1;
    }

    tally[userToken].searchTimes ? tally[userToken].searchTimes.push(searchTime) :
        tally[userToke].searchTimes = searchTime;
```

**Comment by kstaken** [comment]
You should be modeling this as a module that's loaded by search.js. The module should expose an API that search.js can call into to track the various things. So instead of doing things like `res.apiSearchTime = end - start;` you would call something like `stats.timer(end - start)` and then internally to the module it does the aggregation and syncing to ES. 

For tracking counts something like `stats.count(token, responseCode)` or whatever is required.

**Comment by kstaken** [comment]
Something else that just occurred to me is that these stats are going to need to be endpoint specific. The teraserver instance might have 10 different API endpoints configured so the records stored in ES will need to have the endpoint as a field in the record.

**Comment by ciorg** [comment]
If there are multiple workers, then will each worker have it's own count?  I don't always see consecutive counts and I count figure out quite why that happens.  It usually between the first and the second count, I see 1, 1, 2, 3, 4, etc..

**Comment by kstaken** [comment]
Yes each worker will count independently and requests will balance across workers. The analytics viewer will deal with the aggregation of those counts.

**Comment by ciorg** [comment]
Confirmed that each worker has it's own count.  Does it matter overall?  We'll still have an accurate count so long as in ES we aggregate all the records for each api key.

I also am wondering how to capture the errors.  Since the counter is called after performSearch, if there's an error before that the count will never get called and the error status won't be recorded.

And I made it a module loaded into search.js.
